python/pyspark/sql/protobuf/functions.py

```diff
@@ -0,0 +1,196 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""
+A collections of builtin protobuf functions
+"""
+
+
+from typing import Dict, Optional, TYPE_CHECKING
+from pyspark import SparkContext
+from pyspark.sql.column import Column, _to_java_column
+from pyspark.util import _print_missing_jar
+
+if TYPE_CHECKING:
+    from pyspark.sql._typing import ColumnOrName
+
+
+def from_protobuf(
+    data: "ColumnOrName",
+    descFilePath: str,
+    messageName: str,
+    options: Optional[Dict[str, str]] = None,
+) -> Column:
+    """
+    Converts a binary column of Protobuf format into its corresponding catalyst value.
+    The specified schema must match the read data, otherwise the behavior is undefined:
+    it may fail or return arbitrary result.
+    To deserialize the data with a compatible and evolved schema, the expected
+    Protobuf schema can be set via the option protobuf descriptor.
+
+    .. versionadded:: 3.4.0
+
+    Parameters
+    ----------
+    data : :class:`~pyspark.sql.Column` or str
+        the binary column.
+    descFilePath : str
+        the protobuf descriptor in Message GeneratedMessageV3 format.
+    messageName: str
+        the protobuf message name to look for in descriptorFile.
+    options : dict, optional
+        options to control how the Protobuf record is parsed.
+
+    Notes
+    -----
+    Protobuf functionality is provided as an pluggable external module since Spark 3.4.0.
+
+    Examples
+    --------
+    >>> from pyspark.sql import Row
+    >>> import pyspark.sql.types as T
+    >>> from pyspark.sql.protobuf.functions import from_protobuf, to_protobuf
+    >>> data = ([Row(key="1", value=Row(age=2, name="Alice", score=109200))])
+    >>> ddl_schema = "key string, value struct<age: integer, name: string, score: long>"
+    >>> df = spark.createDataFrame(data, T._parse_datatype_string(ddl_schema))
+    >>> descFilePath = 'connector/protobuf/src/test/resources/protobuf/pyspark_test.desc'
```

Comment:
Let's use snake_case unless we have to match it with Java/Scala cases.

BTW, can we just bring this into docstring, and write that down into a temporary file with `with tempfile.TemporaryFile as f`?

We can add an extra comment that how it's generated here. e.g.)

```python
>>> # Write a Protobuf description into a file, generated by blah blah ...
... with tempfile.TemporaryDirectory() as d
... 
```

Reply:
The whole point of doing this is for users to just run the tests by copying and pasting.

Reply:
@HyukjinKwon if I try to initialize .desc file as bytes docstring and then save it to the tmp .desc file. always end up getting malformed protobuf message.

tried the below options. 

```
desc_bytes = b'\n\xce\x01\nAconnector/protobuf/src/test/resources/protobuf/pyspark_test.proto\x12\x1dorg.apache.spark.sql.protobuf"K\n\rSimpleMessage\x12\x10\n\x03age\x18\x01 \x01(\x05R\x03age\x12\x12\n\x04name\x18\x02 \x01(\\tR\x04name\x12\x14\n\x05score\x18\x03 \x01(\x03R\x05scoreB\x15B\x13SimpleMessageProtosb\x06proto3'

or 

desc_bytes = b"\\n\xce\x01\\nAconnector/protobuf/src/test/resources/protobuf/pyspark_test.proto\x12\x1dorg.apache.spark.sql.protobuf\\"K\\n\\rSimpleMessage\x12\x10\\n\x03age\x18\x01 \x01(\x05R\x03age\x12\x12\\n\x04name\x18\x02 \x01(\\tR\x04name\x12\x14\\n\x05score\x18\x03 \x01(\x03R\x05scoreB\x15B\x13SimpleMessageProtosb\x06proto3"

or 

desc_bytes = bytearray("\\n\xce\x01\\nAconnector/protobuf/src/test/resources/protobuf/pyspark_test.proto\x12\x1dorg.apache.spark.sql.protobuf\\"K\\n\\rSimpleMessage\x12\x10\\n\x03age\x18\x01 \x01(\x05R\x03age\x12\x12\\n\x04name\x18\x02 \x01(\\tR\x04name\x12\x14\\n\x05score\x18\x03 \x01(\x03R\x05scoreB\x15B\x13SimpleMessageProtosb\x06proto3".encode())


```
```
While parsing a protocol message, the input ended unexpectedly in the middle of a field.  This could mean either that the input has been truncated or that an embedded message misreported its own length.
```
@rangadi @HyukjinKwon let me know if you know a better way to initialize the bytes and write to tmp .desc file.

Reply:
@HyukjinKwon I had to use Hexa of the protobuf.desc file, which appears to have fewer syntactic errors. Please let me know if you have a better approach.